src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -259,83 +261,93 @@ internal int ReadCore(Span<byte> buffer)
             EnsureDecompressionMode();
             EnsureNotDisposed();
             EnsureBufferInitialized();
-
-            int totalRead = 0;
-
             Debug.Assert(_inflater != null);
+
+            int bytesRead;
             while (true)
             {
-                int bytesRead = _inflater.Inflate(buffer.Slice(totalRead));
-                totalRead += bytesRead;
-                if (totalRead == buffer.Length)
-                {
-                    break;
-                }
-
-                // If the stream is finished then we have a few potential cases here:
-                // 1. DeflateStream => return
-                // 2. GZipStream that is finished but may have an additional GZipStream appended => feed more input
-                // 3. GZipStream that is finished and appended with garbage => return
-                if (_inflater.Finished() && (!_inflater.IsGzipStream() || !_inflater.NeedsInput()))
+                // Try to decompress any data from the inflater into the caller's buffer.
+                // If we're able to decompress any bytes, or if decompression is completed, we're done.
+                bytesRead = _inflater.Inflate(buffer);
+                if (bytesRead != 0 || InflatorIsFinished)
                 {
                     break;
                 }
 
+                // We were unable to decompress any data.  If the inflater needs additional input
```

Comment:
In what case could NeedsInput be false here? Only in the 0-byte read case? Is it even false there? Seems like we should assert about this somehow.

Reply:
> Only in the 0-byte read case?

No.  I'd thought that initially, but this is definitely hit by tests even when the buffer has room available.  I believe it happens at a minimum in situations where there are multiple compressed streams back to back in the input... I think we get here if zlib has hit the end of one stream and stops decompressing even though there's more input available, but you can immediately call back to it to start decompressing the next stream.  I don't know what we could assert about that case, nor if there are others that might get us here.